src/test/ref_tests.cpp

```diff
@@ -0,0 +1,29 @@
+// Copyright (c) 2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <util/ref.h>
+
+#include <boost/test/unit_test.hpp>
+
+BOOST_AUTO_TEST_SUITE(ref_tests)
+
+BOOST_AUTO_TEST_CASE(ref_test)
+{
+    util::Ref ref;
+    BOOST_CHECK(!ref.Has<int>());
+    BOOST_CHECK_THROW(ref.Get<int>(), NonFatalCheckError);
+    int value = 5;
+    ref.Set(value);
+    BOOST_CHECK(ref.Has<int>());
+    BOOST_CHECK_EQUAL(ref.Get<int>(), 5);
+    ++ref.Get<int>();
+    BOOST_CHECK_EQUAL(ref.Get<int>(), 6);
+    BOOST_CHECK(!ref.Has<bool>());
```

Comment:
in the commit that adds ref: Could check that each is an alias for the other? Right now ref could actually be a copy and the test wouldn't fail.

```diff
diff --git a/src/test/ref_tests.cpp b/src/test/ref_tests.cpp
index ef96a0042a..0ec0799fbc 100644
--- a/src/test/ref_tests.cpp
+++ b/src/test/ref_tests.cpp
@@ -19,6 +19,10 @@ BOOST_AUTO_TEST_CASE(ref_test)
     BOOST_CHECK_EQUAL(ref.Get<int>(), 5);
     ++ref.Get<int>();
     BOOST_CHECK_EQUAL(ref.Get<int>(), 6);
+    BOOST_CHECK_EQUAL(value, 6);
+    ++value;
+    BOOST_CHECK_EQUAL(value, 7);
+    BOOST_CHECK_EQUAL(ref.Get<int>(), 7);
     BOOST_CHECK(!ref.Has<bool>());
     BOOST_CHECK_THROW(ref.Get<bool>(), NonFatalCheckError);
     ref.Clear();